### DISCUSSION TITLE
Support for pnpm as an alternative to Yarn

### DISCUSSION THREAD

**Discussion by elliottsj** [vercel/vercel#3896] [Ideas] [answered]
This follows the [previous issue](https://github.com/zeit/now/issues/2973).

pnpm is my preferred package manager because it's fast, saves disk space, and has a strict `node_modules` structure, providing safety guarantees similar to those provided by Yarn PnP. Would love if this can be supported in Now deployments.

**Comment by styfle** [comment]
Hi @elliottsj 

You can use any package manager you wish with `now dev` as long as node can resolve the modules (typically these are located in `node_modules` directory).

From your issue:

> It'll help avoid issues where a website works fine using pnpm locally with now dev, but fails in production after deploying.
> pnpm has a stricter node_modules structure than Yarn, so package resolution behaves differently.

Do you have an example that works with `pnpm install` but fails with `yarn install`?

**Reply by elliottsj** [reply]
Let's say I use pnpm with `now dev`, while the hosted Now deployment continues to use yarn with the same package.json. The problem is that, since I'm using pnpm, I'm not maintaining a `yarn.lock` file. So an unintentionally-breaking change in a transitive dependency can break my application.

This situation might be avoided by _also_ maintaining a `yarn.lock` file; i.e. any time I update dependencies, I'll also run `yarn install` to generate a new `yarn.lock`. But this workaround somewhat negates the improved developer experience of using pnpm alone.

---

Here's an example scenario where using pnpm would save a developer some headache:

1. I'm using next@9.3.1 and Yarn v1.
2. I decide to import an npm package that happens to be a transitive dependency of next@9.3.1, e.g. [chalk@2.4.2](https://github.com/chalk/chalk/tree/v2.4.2): https://github.com/zeit/next.js/blob/d3fb2629585839da2f70eb5bdd88f47848a7d582/packages/next/package.json#L88

   ```ts
   import chalk from 'chalk';
   ```

3. I forget to add `"chalk"` to my package.json, but the above code still works because Yarn's package hoisting makes the package available at `node_modules/chalk/`.
4. Some time later, next@9.4.0 is released and now depends on chalk@3.0.0.
5. My app breaks because it was written for chalk@2.4.2, but now `node_modules/chalk/` is on v3.0.0.

Using pnpm, the problem would be caught at step (3), because pnpm does not hoist transitive dependencies to the root `node_modules` folder.

**Reply by styfle** [reply]
Interesting use case!

I tried to reproduce, but step 3 works the same with both `yarn` and `pnpm`:

### yarn
```
mkdir yarn-example
cd yarn-example
yarn init -y
yarn add next
echo "console.log(require('chalk'))" > index.js
echo "console.log(require('path').resolve('chalk'))" > path.js
node index.js # chalk was found
node path.js # chalk was found
```

### pnpm
```
mkdir pnpm-example
cd pnpm-example
yarn init -y
pnpm add next
echo "console.log(require('chalk'))" > index.js
echo "console.log(require('path').resolve('chalk'))" > path.js
node index.js # chalk was found
node path.js # chalk was found
```

**Reply by elliottsj** [reply]
Oh, that's strange 🤔 

Can you try your pnpm example again? This is my result:

<img width="934" alt="Screen Shot 2020-03-23 at 12 39 31 PM" src="https://user-images.githubusercontent.com/2177366/77340338-890ecf00-6d03-11ea-9f92-9ce3f7340936.png">

**Reply by styfle** [reply]
I see now. I had node_modules in a folder above so it was resolving.

I confirmed that running `require('chalk')` after pnpm install will throw and it will not after yarn install.

I suppose your workaround today would be to change your Build Script to `rm -rf node_modules && npx pnpm install`.

**Reply by elliottsj** [reply]
Here is my current workaround:

#### package.json

```json
"scripts": {
  "postinstall": "node -e \"process.exitCode = process.env.npm_execpath.includes('pnpm') ? 1 : 0;\" && npx pnpm@4.13.0 install || echo \"dependencies installed with pnpm\"",
}
```

**Comment by remorses** [comment]
Hi, would adding pnpm support to this function resolve the issue?

https://github.com/vercel/vercel/blob/e2db7c77347fe7698a76df8f31062b9c3b1490b4/packages/now-build-utils/src/fs/run-user-scripts.ts#L262

I am not sure if you have non public code that would need to be changed to support pnpm, if that is not the case i can open a PR.

**Reply by paulogdm** [reply]
We are working on implementing a field that you can insert a custom install command. PNPM will be supported when that happens.

**Reply by mazipan** [reply]
Is there any progress tracker that we can follow for this request @paulogdm ?

**Reply by paulogdm** [reply]
Not yet. Should be here in the next month if everything goes right.

**Reply by intrnl** [reply]
I hope that it supports workspaces scenario as well :joy:

**Reply by danvoyce** [reply]
Any updates on this. I'd love to switch over to Vercel, but this is a blocker for me

**Comment by danvoyce** [comment]
I've managed to wriggle around this issue by doing the following in the `install command` field...

```
npm i pnpm -g && pnpm i
```

**Reply by styfle** [reply]
Indeed, you can select "Override" and specify the desired [Install Command](https://vercel.com/docs/build-step#install-command) in your project settings.

**Reply by elliottsj** [reply]
I'm trying the new install command override, setting it to `npx pnpm install`.

However, new builds seem to error with:

```
TypeError: _chalk.default.constructor is not a constructor
    at getChalk (/vercel/workpath0/node_modules/.pnpm/@babel/highlight@7.10.4/node_modules/@babel/highlight/lib/index.js:90:13)
    at codeFrameColumns (/vercel/workpath0/node_modules/.pnpm/@babel/code-frame@7.10.4/node_modules/@babel/code-frame/lib/index.js:92:41)
    at getFormattedDiagnostic (/vercel/workpath0/node_modules/.pnpm/next@10.0.1_673d10ab695550cf3e856d21663dfc3f/node_modules/next/dist/lib/typescript/diagnosticFormatter.js:4:687)
    at runTypeCheck (/vercel/workpath0/node_modules/.pnpm/next@10.0.1_673d10ab695550cf3e856d21663dfc3f/node_modules/next/dist/lib/typescript/runTypeCheck.js:1:1436)
    at async verifyTypeScriptSetup (/vercel/workpath0/node_modules/.pnpm/next@10.0.1_673d10ab695550cf3e856d21663dfc3f/node_modules/next/dist/lib/verifyTypeScriptSetup.js:8:8)
    at async build (/vercel/workpath0/node_modules/.pnpm/next@10.0.1_673d10ab695550cf3e856d21663dfc3f/node_modules/next/dist/build/index.js:4:960)
```

Seems to happen on builds that restore the build cache from the previous build. I suspect it's due to Vercel persisting/restoring pnpm's node_modules store incorrectly.

**Reply by styfle** [reply]
If you add `rm -rf node_modules` before pnpm install does that fix it?

Can you provide an example repo that fails?

Thanks!

**Reply by elliottsj** [reply]
This is my repo: https://github.com/elliottsj/elliott.dev/

Works after changing the override to `rm -rf node_modules && npx pnpm install` 👍

**Reply by styfle** [reply]
Try `npx pnpm i` instead.


If you had just updated the install command, it would have used the old cache from npm/yarn which might have been the problem.

I am not able to reproduce the issue with your repo, even after subsequent deploys so it seems like it was just outdated cache.

**Reply by intrnl** [reply]
Would it still be a problem if the store was moved to inside of node_modules?

`pnpm install --store=node_modules/.pnpm-store`

This is something that I do for CI to cache both the node_modules and the store itself usually.

**Reply by elliottsj** [reply]
Tried `npx pnpm install --store=node_modules/.pnpm-store`; unfortunately same result: https://gist.github.com/elliottsj/a964ec439763687bd50b1197ef70404f

**Reply by elliottsj** [reply]
A more complete repro: https://github.com/elliottsj/nextjs-vercel-pnpm-issue

**Reply by styfle** [reply]
Thanks, this bug is being tracked here: https://github.com/vercel/next.js/issues/16471

**Reply by elliottsj** [reply]
https://github.com/vercel/next.js/issues/16471 was closed but this issue is not yet fixed. Should I create a new issue?

**Reply by styfle** [reply]
Sure, if you can still reproduce the Next.js pnpm issue then create a new bug, thanks 👍

**Reply by elliottsj** [reply]
Done: https://github.com/vercel/next.js/issues/20948
Proposed fix: https://github.com/vercel/next.js/pull/20949

**Reply by kidonng** [reply]
It seems this doesn't apply to serverless functions. Vercel will ignore this setting and use Yarn because no `package-lock.json` is found.

**Reply by danvoyce** [reply]
This eventually has bit me in the a$$... https://github.com/vercel/next.js/issues/16471#issuecomment-841212671

**Reply by hyoretsu** [reply]
Is there a way to automate it and not edit Install field on Vercel for every site?

**Reply by bodadotsh** [reply]
Don't forget to change the build step to `pnpm build` else it will still run `yarn run build`

**Reply by intrnl** [reply]
Does Vercel's env configuration apply to the entire build environment?

npm allows passing command line arguments via `NPM_FLAGS` environment variable and I believe Yarn also has this too via `YARN_FLAGS`.

We can use this to our advantage to skip the unnecessary install procedure and go straight to pnpm's install command, which could be faster.

`NPM_FLAGS=--version`

**Reply by privatenumber** [reply]
As a shortcut, you can also use the command:

```
npx ci
```


It will automatically toggle between npm, pnpm, and yarn depending on the lock file.

**Comment by bodadotsh** [comment, marked as answer]
https://vercel.com/changelog/projects-using-pnpm-can-now-be-deployed-with-zero-configuration

Great to see Vercel supports PNPM natively now!
